src/nccl_ofi_idpool.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "nccl_ofi.h"
+#include "nccl_ofi_idpool.h"
+#include "nccl_ofi_math.h"
+
+/*
+ * @brief	Initialize pool of IDs
+ *
+ * Allocates and initializes a nccl_ofi_idpool_t object, marking all
+ * IDs as available.
+ *
+ * @param	idpool_p
+ *		Return value with the ID pool pointer allocated
+ * @param	size
+ *		Size of the id pool (number of IDs)
+ * @return	0 on success
+ *		non-zero on error
+ */
+int nccl_ofi_idpool_init(nccl_ofi_idpool_t *idpool, size_t size)
+{
+	int ret = 0;
+
+	assert(NULL != idpool);
+
+	if (0 == size) {
+		/* Empty or unused pool */
+		idpool->ids = NULL;
+		idpool->size = 0;
+		return ret;
+	}
+
+	/* Scale pool size to number of 64-bit uints (rounded up) */
+	size_t num_long_elements = NCCL_OFI_ROUND_UP(size, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
+
+	/* Allocate memory for the pool */
+	idpool->ids = malloc(sizeof(uint64_t) * num_long_elements);
+
+	/* Return in case of allocation error */
+	if (NULL == idpool->ids) {
+		NCCL_OFI_WARN("Unable to allocate ID pool");
+		return -ENOMEM;
+	}
+
+	/* Set all IDs to be available */
+	memset(idpool->ids, 0xff, size / 8);
+	if (size % 8) {
+		idpool->ids[num_long_elements - 1] = (1ULL << (size % (sizeof(uint64_t) * 8))) - 1;
+	}
+
+	/* Initialize mutex */
+	ret = pthread_mutex_init(&idpool->lock, NULL);
+	if (OFI_UNLIKELY(ret)) {
+		NCCL_OFI_WARN("Unable to initialize mutex");
+		free(idpool->ids);
+		idpool->ids = NULL;
+		return ret;
+	}
+
+	idpool->size = size;
+
+	return ret;
+}
+
+/*
+ * @brief	Allocate an ID
+ *
+ * Extract an available ID from the ID pool, mark the ID as
+ * unavailable in the pool, and return extracted ID. No-op in case
+ * no ID was available.
+ *
+ * This operation is locked by the ID pool's internal lock.
+ *
+ * @param	idpool
+ *		The ID pool
+ * @return	the extracted ID (zero-based) on success,
+ *		negative value on error
+ */
+int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool)
+{
+	assert(NULL != idpool);
+
+	if (0 == idpool->size) {
+		NCCL_OFI_WARN("Cannot allocate an ID from a 0-sized pool");
+		return -ENOMEM;
+	}
+
+	if (OFI_UNLIKELY(NULL == idpool->ids)) {
+		NCCL_OFI_WARN("Invalid call to nccl_ofi_allocate_id with uninitialized pool");
+		return -EINVAL;
+	}
+
+	/* Scale pool size to number of 64-bit uints (rounded up) */
+	size_t num_long_elements = NCCL_OFI_ROUND_UP(idpool->size, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
```

Comment:
This is a really expensive operation for the critical path.  I don't understand why we need it; just handle as part of a loop from 0 ... idpool->size.  `sizeof(uint64_t) * 8` is going to be a power of 2, so `%` will be a cheap operation.